martin-mbtiles/src/tile_copier.rs

```diff
@@ -144,207 +145,278 @@ impl TileCopier {
-                             ON sourceDb.tiles.zoom_level = newDb.tiles.zoom_level
-                             AND sourceDb.tiles.tile_column = newDb.tiles.tile_column
-                             AND sourceDb.tiles.tile_row = newDb.tiles.tile_row
-                        WHERE (sourceDb.tiles.tile_data != newDb.tiles.tile_data
-                            OR sourceDb.tiles.tile_data ISNULL
-                            OR newDb.tiles.tile_data ISNULL)",
-            )
-            .await
-        } else {
-            self.run_query_with_options(
-                conn,
-                // Allows for adding clauses to query using "AND"
-                &format!(
-                    "INSERT {} INTO tiles SELECT * FROM sourceDb.tiles WHERE TRUE",
-                    match &self.options.on_duplicate {
-                        CopyDuplicateMode::Override => "OR REPLACE",
-                        CopyDuplicateMode::Ignore | CopyDuplicateMode::Abort => "OR IGNORE",
-                    }
-                ),
-            )
-            .await
+    fn get_on_duplicate_sql(&self, mbttype: &MbtType) -> (String, String) {
+        match &self.options.on_duplicate {
+            CopyDuplicateMode::Override => ("OR REPLACE".to_string(), "".to_string()),
+            CopyDuplicateMode::Ignore => ("OR IGNORE".to_string(), "".to_string()),
+            CopyDuplicateMode::Abort => ("OR ABORT".to_string(), {
+                let (main_table, tile_identifier) = match mbttype {
+                    Flat => ("tiles", "tile_data"),
+                    FlatHashed => ("hashed_tiles", "tile_data"),
+                    Normalized => ("map", "tile_id"),
+                };
+
+                format!(
+                        "AND NOT EXISTS (\
+                        SELECT 1 \
+                        FROM {main_table} \
+                        WHERE \
+                            {main_table}.zoom_level=sourceDb.{main_table}.zoom_level \
+                            AND {main_table}.tile_column=sourceDb.{main_table}.tile_column \
+                            AND {main_table}.tile_row=sourceDb.{main_table}.tile_row \
+                            AND {main_table}.{tile_identifier}!=sourceDb.{main_table}.{tile_identifier}\
+                        )"
+                    )
+            }),
         }
     }
 
-    async fn copy_deduplicated(&self, conn: &mut SqliteConnection) -> MbtResult<()> {
-        let on_duplicate_sql = match &self.options.on_duplicate {
-            CopyDuplicateMode::Override => "OR REPLACE",
-            CopyDuplicateMode::Ignore | CopyDuplicateMode::Abort => "OR IGNORE",
+    fn get_select_from_with_diff(&self, dst_mbttype: &MbtType, diff_mbttype: &MbtType) -> String {
+        let (hash_col_sql, new_tiles_with_hash) = if dst_mbttype == &Flat {
+            ("", "newDb.tiles")
+        } else {
+            match *diff_mbttype {
+                Flat => (", hex(sha256(tile_data)) as hash", "newDb.tiles"),
```

Comment:
for now, lets use `md5` everywhere - simply because most other implementations use md5. This is **not** a good hash though, so in the future (another PR) we should introduce an option to use sha256